GNUmakefile

```diff
@@ -1,77 +1,158 @@
-		exit 1; \
-		fi
-	@sh -c "'$(PWD)/scripts/test.sh'"
-	@$(MAKE) vet
-
-cover:
-	go list ./... | xargs -n1 go test --cover
-
-format:
-	@echo "--> Running go fmt"
-	@go fmt $(PACKAGES)
-
-generate:
-	@echo "--> Running go generate"
-	@go generate $(PACKAGES)
-	@sed -i.old -e 's|github.com/hashicorp/nomad/vendor/github.com/ugorji/go/codec|github.com/ugorji/go/codec|' nomad/structs/structs.generated.go
-
-vet:
-	@echo "--> Running go vet $(VETARGS) ${PACKAGES}"
-	@go vet $(VETARGS) ${PACKAGES} ; if [ $$? -eq 1 ]; then \
-		echo ""; \
-		echo "[LINT] Vet found suspicious constructs. Please check the reported constructs"; \
-		echo "and fix them if necessary before submitting the code for review."; \
-		exit 1; \
-		fi
-	@git grep -n `echo "log"".Print"` | grep -v 'vendor/' ; if [ $$? -eq 0 ]; then \
-		echo "[LINT] Found "log"".Printf" calls. These should use Nomad's logger instead."; \
-		fi
-
-# bootstrap the build by downloading additional tools
-bootstrap:
-	@for tool in  $(EXTERNAL_TOOLS) ; do \
-		echo "Installing $$tool" ; \
-		go get $$tool; \
-		done
-	@for tool in  $(TEST_TOOLS) ; do \
-		echo "Installing $$tool (test dependency)" ; \
-		go get $$tool; \
-		done
-
-install: bin/nomad
-	install -o root -g wheel -m 0755 ./bin/nomad /usr/local/bin/nomad
-
-travis:
-	@sh -c "'$(PWD)/scripts/travis.sh'"
-
-.PHONY: all bin cov integ test vet test-nodep
+PROJECT_ROOT := $(patsubst %/,%,$(dir $(abspath $(lastword $(MAKEFILE_LIST)))))
+THIS_OS := $(shell uname)
+
+GIT_COMMIT := $(shell git rev-parse HEAD)
+GIT_DIRTY := $(if $(shell git status --porcelain),+CHANGES)
+GO_LDFLAGS := "-X main.GitCommit=$(GIT_COMMIT)$(GIT_DIRTY)"
+GO_TAGS =
+
+GOOS := $(shell go env GOOS)
+GOARCH := $(shell go env GOARCH)
+GOPATH := $(shell go env GOPATH)
+
+LOCAL_PACKAGES = $(shell go list ./... | grep -v '/vendor/')
+
+# Enable additional linters as the codebase evolves to pass them
+CHECKS ?= --enable goimports
+
+ifeq (,$(findstring $(THIS_OS),Darwin Linux))
+$(error Building Nomad is currently only supported on Darwin and Linux.)
+endif
+
+default: help
+
+# Targets are dynamically constructed below
+ALL_TARGETS =
+ALL_PACKAGES =
+
+define buildNomad
+# 1: GOOS
+# 2: GOARCH
+# 3: CGO_ENABLED
+# 4: TAGS
+# 5: CC
+# 6: CXX
+
+ALL_TARGETS += $(1)_$(2)$(if $(4),-$(4))
+ALL_PACKAGES += pkg/$(1)_$(2)$(if $(4),-$(4)).zip
+
+pkg/$(1)_$(2)$(if $(4),-$(4))/nomad: $(SOURCE_FILES)
+	@echo "==> Building for $(1)/$(2)$(if $(4), with $(4))..."
+	@CGO_ENABLED=$(3) \
+		GOOS="$(1)" \
+		GOARCH="$(2)" \
+		$(if $(5),CC="$(5)") \
+		$(if $(6),CXX="$(6)") \
+		go build \
+		-ldflags $(GO_LDFLAGS) \
+		-tags "$(GO_TAGS) $(4)" \
+		-o "pkg/$(1)_$(2)$(if $(4),-$(4))/nomad"
+
+pkg/$(1)_$(2)$(if $(4),-$(4)).zip: pkg/$(1)_$(2)$(if $(4),-$(4))/nomad
+	@echo "==> Packaging for $(1)/$(2)$(if $(4), with $(4))..."
+	@zip pkg/$(1)_$(2)$(if $(4),-$(4)).zip pkg/$(1)_$(2)$(if $(4),-$(4))/*
+endef
+
+# We build only Darwin binaries on Mac OS X
```

Comment:
This is what defines the Darwin target, so it needs to be at the top level. However, it would likely be clearer if the target template function didn't actual building action: `buildNomad` should be something like `definePlatformBuildTarget`. I'll rename that, rebase and push a new version.